### ISSUE TITLE
Floccus hangs

### ISSUE THREAD

**Issue by mobamoba** [marcelklehr/floccus#73]
### Software versions

* Browser: Firefox
* Nextcloud: 13
* Nextcloud Bookmarks app: most recent
* Floccus: 2.05

### Steps to reproduce

1. Nothing.
2. Floccus hangs.

### Expected outcome

...

### Actual outcome

...

Floccus hangs after some period of time. The syncing symbol appears on the icon, but clicking it brings up nothing. The Options menu also brings up nothing other than the main description page. Here's the debug log:

Error: Could not establish connection. Receiving end does not exist. undefined
Error: Invalid tab ID: 119  ext-browser.js
Error: Invalid tab ID: 122 ext-browser.js
Error: Could not establish connection. Receiving end does not exist. undefined
Unknown property ‘enable-background’.  Declaration dropped.  container-edit.svg
Unknown property ‘enable-background’.  Declaration dropped.  container-delete.svg
Error: Could not establish connection. Receiving end does not exist. undefined
TypeError: can't access dead object[Learn More]  core.js:682:17
Error: Could not establish connection. Receiving end does not exist.  undefined

**Comment by marcelklehr** [comment]
Mmmh. Any CPU load? Does disabling and enabling the addon help?

**Comment by mobamoba** [comment]
Yeah there was some CPU load but with Firefox it's difficult to tie to a particular addon. And disabling/enabling didn't help.

**Comment by mobamoba** [comment]
I've narrowed this down a bit. Floccus is incompatible with Firefox Sync for some reason even when sync isn't syncing Bookmarks. In other words, having Sync turned on, even if it's doing nothing other than syncing open browser tabs, will at some point hang Floccus. When Sync gets stuck in a loop of some kind (which it seems to quite often), Floccus hangs.

I'm not sure why these two things are connected as all other addons work fine regardless of what Sync is doing but that's pretty much what I've narrowed it down to.

**Comment by marcelklehr** [comment]
It could be that Sync temporarily stops some API interactions. If it get stuck in a loop as you described, the bookmarks API will be blocked forever and so will floccus. I recommend using either Sync or Floccus even though I realize that floccus doesn't offer the full set of features of Firefox Sync. :/

(Closing this for now. Feel free to reopen!)

**Comment by mobamoba** [comment]
I think we should reopen. Firefox sync is disabled and Firefox restarted, Floccus is up to 2.0.6. Same problem. And unfortunately same lack of debugging information,

**Comment by marcelklehr** [comment]
Do you have a lot of bookmarks? Maybe it's not actually hanging, but lagging extremely. I've experienced poor performance in Firefox before.

**Comment by mobamoba** [comment]
Not really - 700 or so? I don't think it's lag. I've just left it off hoping that maybe a new version would magically fix the problem. :)

**Comment by dotancohen** [comment]
I'm experiencing a similar issue, but I have what some people would call a lot of bookmarks.
```
mysql> select count(*) from oc_bookmarks;
|     3808 |
```

Firefox hangs often since installing Floccus.
Clicking the Floccus icon just opens an empty bubble.
Adding a new bookmark to the sync folder freezes up my entire system for about half a minute.
The options page sometimes shows my credentials, and sometimes shows only the main description text.
I've tried disabling all other addons (Tridactyl, Tree Style Tabs) but that did not help.
Firefox Sync is disabled, I do not have a Firefox account logged in to the browser.

CentOS Linux 7.3, Firefox 59.0.2 downloaded from Mozilla.

I'm happy to help if I can provide any more information.

**Comment by marcelklehr** [comment]
> Clicking the Floccus icon just opens an empty bubble.
Does this happen even with the latest version (v2.1.0)?

> Adding a new bookmark to the sync folder freezes up my entire system for about half a minute.
Same here. It shouldn't freeze immediately in the latest version, but only after a timeout of inactivity.

**Comment by dotancohen** [comment]
![image](https://user-images.githubusercontent.com/246144/39530299-5c8709fe-4e31-11e8-9d6e-c7f0b2862721.png)

**Comment by dotancohen** [comment]
![image](https://user-images.githubusercontent.com/246144/39530356-781eddb8-4e31-11e8-9b52-74edf2f27189.png)

**Comment by dotancohen** [comment]
![image](https://user-images.githubusercontent.com/246144/39530434-a2671ed2-4e31-11e8-9084-a4dd9ea75c7f.png)

**Comment by exekutive** [comment]
I have the same symptom as dotancohen. Just a blank bubble.

Nextcloud 13.0.2
Bookmarks app 0.11.0
floccus 2.1.0
Firefox sync disabled

**Comment by marcelklehr** [comment]
Please try v2.2.0 to check whether this is still an issue in the latest version.

**Comment by dotancohen** [comment]
Thank you, it does seem that v2.2.0 works fine and is not hanging.
Note that the hang may have been due to one of a few issues which I've identified:

1. I've changed by MC URL from http to https, and though I updated that in the Floccus settings, it seems to not have been saved in Floccus. I could not know that with 2.1, as the UI for the server config would often be not displayed. With v2.2 I saw the issue and immediately corrected it.

2. I actually have many more bookmarks than previously stated. Though I found ~4000 bookmarks synced to NC, in places.sqlite I see quite a few more, but v2.2 is handling them fine:
```
sqlite> select count(*) from moz_bookmarks;
26455
```

3. I have a few bookmarks that are dupes in different folders. There is no way around this, I have a separate folder for each client, each which has a sub-sub-sub folder for their project dependencies' homepages. As I may use i.e. Python Requests for multiple projects, so must I have the Requests homepage in multiple folders. However, other than the stream of error messages Floccus seems to handle this situation fine. I'm waiting anxiously for folder support in Floccus!

**Comment by dotancohen** [comment]
I spoke too soon. After about an hour of syncing, during which ~9000 bookmarks have been transfered to the server, Firefox is using a bit over 100% CPU (8 core CentOS desktop) and over 2 GiB of RAM (with electrolosys - multiprocess firefox - it is difficult to say exactly). The whole system is sluggish. With another ~10,000 bookmarks to go I'm afraid of stability and there seems to be no way to cancel a running sync.

**Comment by marcelklehr** [comment]
Mh. I'd say it wasn't designed to sync ~30000 bookmarks :D But then it's slow even for 3000 which is my current testing benchmark. I'll investigate.

Your miraculous multiplication of bookmarks could be due to some bookmarks being stuck in the root folder of your browser, where you can't see them. :/

**Comment by marcelklehr** [comment]
Am I correct in assuming that the "empty bubble" problem is gone now?

**Comment by dotancohen** [comment]
Reverting to 2.2.0 and restarting Firefox did not resolve the issue. So it is _not_ a problem with 2.2.2, and I'm returning to that version.

**Comment by dotancohen** [comment]
Thank you, here are the logs:
https://pastebin.com/6E2Q4JVg
Note that the fake "foo" account shows repeatedly, yet the real Nextcloud account does not show at all.

**Comment by dotancohen** [comment]
I have now completely removed Floccus, restarted Firefox, and reinstalled 2.2.2 from AMO. I'll report back with the results.

**Comment by dotancohen** [comment]
With the new install:
1. The empty bubble is gone, I have a proper menu when clicking the toolbar item.
2. I see some _slow_ syncing happening. Five bookmarks were added to the database on the NC server within five minutes of syncing.
3. I have two foo@http://ba.r accounts created, even though I was careful to select my proper folder for syncing and _not_ to create the "bob" account or press "test" button (I don't even see it anymore).
4. The system sluggeshness when syncing seems to be resolved, at least for the first five minutes of syncing. I'll let it run all night, I'm leaving the office now, and I'll report back in the morning.

![selection_001](https://user-images.githubusercontent.com/246144/40019965-70912690-57c9-11e8-8d95-093a79035c04.png)

**Comment by marcelklehr** [comment]
> I have two foo@http://ba.r accounts created, even though I was careful to select my proper folder for syncing and not to create the "bob" account or press "test" button (I don't even see it anymore).

What do you mean by "created"? Are they visible in the floccus options panel or do you only see the folders? The folders you can safely delete. (The former shouldn't be possible at all, since those active dummy accounts were the reason you got the empty bubble in the first place :) )

**Comment by marcelklehr** [comment]
> The system sluggeshness when syncing seems to be resolved, at least for the first five minutes of syncing. I'll let it run all night, I'm leaving the office now, and I'll report back in the morning.
The first phase of the sync only stresses the server. The heavy lifting for your firefox instance comes at the end ;)

**Comment by dotancohen** [comment]
> What do you mean by "created"?

I mean that I see their folders in the Bookmarks, nothing else. I should have been clearer.

After letting Floccus run all night, I see that I have 23589 bookmarks on the server. That is about 3000 bookmarks short of what Firefox is holding, but I understand that Floccus with not store dupes as it does not support folders. 


> The first phase of the sync only stresses the server. The heavy lifting for your firefox instance comes at the end ;)

Server:
```
--------        CPU     %user     %nice   %system   %iowait    %steal     %idle
19:25:01        all      0.53      0.00      0.22      0.31      0.11     98.83
19:35:02        all      0.55      0.00      0.20      0.36      0.14     98.75
19:45:01        all      3.34      0.00      0.26      0.03      0.07     96.29
19:55:01        all     93.41      0.00      1.72      2.26      0.31      2.30
20:05:01        all     93.94      0.00      1.68      2.61      0.64      1.14
20:15:01        all     95.50      0.00      1.69      1.79      0.42      0.60
20:25:01        all     94.62      0.00      1.70      2.48      0.26      0.93
20:35:01        all     94.44      0.00      1.75      2.35      0.72      0.74
20:45:01        all     82.31      0.00      1.66      2.95      0.46     12.62
20:55:01        all     67.28      0.00      1.49      4.49      0.48     26.26
21:05:01        all     59.75      0.00      1.41      4.97      0.54     33.33
21:15:02        all     55.97      0.00      1.28      3.52      0.52     38.71
21:25:01        all     52.36      0.00      1.23      5.77      0.34     40.31
21:35:01        all     48.81      0.00      1.16      4.34      0.64     45.05
21:45:01        all     46.58      0.00      1.15      4.32      0.50     47.45
21:55:02        all     18.26      0.00      0.58      2.57      0.25     78.34
22:05:01        all      0.64      0.00      0.30      0.63      0.08     98.34
22:15:01        all      0.50      0.00      0.30      0.65      0.08     98.48
22:25:01        all      0.36      0.00      0.19      0.11      0.18     99.16
22:35:01        all      0.54      0.00      0.26      0.32      0.17     98.71
```

Desktop:
```
20:00:01        CPU     %user     %nice   %system   %iowait    %steal     %idle
20:10:01        all      6.65      0.00      0.78      0.31      0.00     92.26
20:20:01        all      8.60      0.00      0.88      0.27      0.00     90.25
20:30:01        all     10.96      0.00      1.05      0.23      0.00     87.75
20:40:01        all     12.61      0.01      1.14      0.27      0.00     85.97
20:50:01        all     12.87      0.00      1.14      0.22      0.00     85.77
21:10:01        all     12.84      0.00      1.21      0.22      0.00     85.73
21:20:01        all     12.83      0.00      1.16      0.20      0.00     85.80
21:30:01        all     12.80      0.00      1.15      0.19      0.00     85.86
21:40:01        all     12.83      0.01      1.15      0.20      0.00     85.82
21:50:01        all     12.75      0.00      1.15      0.21      0.00     85.90
22:00:01        all      6.21      0.00      0.72      0.20      0.00     92.87
22:10:01        all      1.16      0.01      0.35      0.28      0.00     98.21
22:20:01        all      1.44      0.01      0.37      0.22      0.00     97.96
22:30:01        all      1.42      0.00      0.37      0.21      0.00     98.00
22:40:01        all      0.87      0.08      0.34      0.20      0.00     98.52
```

I converted all times to UTC so that they could be compared. The server is an AWS T2.small, but the accrued CPU burst credits seem to have been enough.

**Comment by marcelklehr** [comment]
If your browser is not in %user then this still doesn't look good, but as expected, per %system. If the %system burst is not from your browser, then this looks remarkably good.

**Comment by marcelklehr** [comment]
@mobamoba Any news on your end?

**Comment by mobamoba** [comment]
For me things have been working perfectly for the last few versions. At some point, I started a new Firefox profile and that solved the problem I think.

**Comment by dotancohen** [comment]
Yes, of course the browser is in %user. Note that the browser is on the Desktop, which only hit ~12.
